lib/Optimizer/GraphOptimizer.cpp

```diff
@@ -855,6 +856,81 @@ static void optimizeBatchNorm(Function *F) {
   } // For all nodes in the graph.
 }
 
+/// Check if the input of this method (vector of NodeValues) can be
+/// legal input of CreateConcat function.
+/// \returns the dimension where the Concat will take place. If not
+/// concatenatable, return -1.
+static int getConcatDimensionForInputs(std::vector<NodeValue> inputs,
+                                       ConcatNode *origCN) {
+  auto firstDims = inputs.front().dims();
+  int dimSize = firstDims.size();
+  auto origCNInputShape = origCN->getInputs().front().dims();
+  int origCNDim = origCN->getDim();
+  int minimalNumOfUnitToLegallyConcat = 1;
+  // Find the minimal number of data units so that we can legally Concat them.
+  for (int i = origCNInputShape.size() - 1; i >= origCNDim; --i) {
+    minimalNumOfUnitToLegallyConcat *= origCNInputShape[i];
+  }
+
+  int index = -1;
+  int count = 0;
+  int preIndex = -1;
+  for (auto &in : inputs) {
+    auto curDims = in.dims();
+    // Find the previous non-negative index value.
+    preIndex = index == -1 ? preIndex : index;
+    count = 0;
+    index = -1;
+    for (int i = 0; i < dimSize; ++i) {
+      // Find the index i where firstDims[i] != curDims[i]
+      if (firstDims[i] != curDims[i]) {
+        count += 1;
+        index = i;
+      }
+    }
+    if (count > 1) {
+      // If curDims and firstDims have more than one different element ,
+      // then we can not perform Concat on top of these two nodes.
+      return -1;
+    }
+    // We cannot have two different non-negative index to be the
+    // dim for Concat.
+    if (index != -1 && preIndex != -1 && index != preIndex) {
+      return -1;
+    }
+  }
+
+  // Determine the index that can be the dim for Concat.
+  index = index == -1 ? preIndex : index;
+  // It's still possible that index can be -1. it means all inputs have
+  // same shape.
+  if (index == -1) {
+    // In this case, we find the index that can meet the requirement of
+    // minimal number of data units to legally concat.
+    int cumulativeNumUnits = 1;
+    for (int i = dimSize - 1; i >= 0; i--) {
+      cumulativeNumUnits *= firstDims[i];
+      // Go over the dimensions from left to right, find the first index that
```

Comment:
Isn't this from right to left, since you're reverse iterating from `i = dimSize - 1 to 0`?

Reply:
changed. Thx